Separate brand model

shuesken opened the pull request.

shuesken commented:
I removed the inheritance of `Datasource` from `Brand`. This includes removing the `source_link` field from `Brand`, adding `name` and `source_link` field to `Datasource`, and adding the relevant fields per `Datasource` subclass. I was considering using mixins or some abstract class to avoid repeating fields like `website` and `description` all the time, but decided it was sacrificing too much code readability.

In the process of adapting the `Datasource` subclasses, I also removed the `tag` field and related code from all of them.

I left the countries fields, subsidiary relationships, and brand-datasource associations untouched at this point too keep the diff a bit more manageable.

RogerTangos commented on datasource/models/wikidata/wikidata.py:
```diff
@@ -1,5 +1,8 @@
 import traceback
```
sorry - i should have noticed this before, but would it be possible to migrate wikidata.py to the top of datasource/model and then move the other files into datasource/local/wikidata? That's the pattern the other datasources are following.

If you feel strongly about it, it would also be possible to move each of the datasources into their own subfolder in datasource/model, but I'd like to keep it consistent

shuesken replied:
done

RogerTangos commented on datasource/models/wikidata/wikidata.py:
```diff
@@ -53,17 +52,14 @@ def only_allowed(val):
-                num_created, existing_tags = cls._maybe_create_individual_instance(
-                    existing_tags, banks, num_created, bank_df
-                )
+                num_created = cls._maybe_create_individual_instance(banks, num_created, bank_df)
```
ok. yeah, I can see how not dealing with tags makes life way easier. Looks like you forgot to update the tests for this though.

shuesken replied:
fixed tests

RogerTangos commented on datasource/models/wikidata/wikidata.py:
```diff
@@ -205,10 +180,29 @@ def _maybe_create_individual_instance(cls, existing_tags, banks, num_created, df
-        return num_created, existing_tags
+        return num_created
+
+    permid = models.CharField(max_length=15, blank=True)
```
As long as you're adding these here, why not also add aliases? I believe that wikidata has those.

shuesken replied:
I didn't add any fields beyond what was need to maintain current functionality. Happy to make changes (here and to Banktrack) to this PR as well, but figured that's a more atomic change for the future.

RogerTangos commented on brand/models/brand.py:
```diff
@@ -195,21 +189,6 @@ def refresh(self, name=True, description=True, countries=True, overwrite_existin
         if countries:
             self.refresh_countries()
 
-    def subclass(self):
```
This should and the related tests should be migrated to datasource.py. It's more useful than you might expect.

shuesken replied:
done; created a test of it in BanktrackTestCase

RogerTangos commented on datasource/models/banktrack.py:
```diff
@@ -97,3 +96,24 @@ def _generate_tag(cls, bt_tag, increment=0, existing_tags=None):
             return bt_tag
         else:
             return cls._generate_tag(og_tag, increment=increment + 1, existing_tags=existing_tags)
+
```
No action required on this one, but FYI: The Banktrack API is giving us much, much richer information than we are currently saving. It may be worth saving that in the future.

for example:
- What projects a bank is funding, along with their years, categorical information, amount(s) and a human-written description

shuesken replied:
see above Wikidata: happy to do that, but I think better on a future PR

RogerTangos commented on datasource/models/wikidata/wikidata.py:
```diff
@@ -72,7 +68,7 @@ def only_allowed(val):
                 traceback.print_exc()
 
         def add_subsidiary(bank, parent):
```
Won't this break? Subsidiary information remains in the Brand model but now wikidata doesn't inherit from brand. I'd suggest either commenting this out to deal with it later or building a binary subsidiary model into wikidata. (i.e. `subsidiary_of=models.ForeignKey(Wikidata...)`

shuesken replied:
yes this would break; I'd removed the code that calls it but now also commented this out.

RogerTangos reviewed (changes requested):
Thanks for the big PR!

Generally, I like where it’s going, though there are still some things to be done.

- Would it also make sense to also have brand inherit from TimeStampedModel? Doing that with Datasource was smart.
- remember to run and update the tests! `django test datasource` or just `django test`
- please manually and adjust ingest_airtable and refresh_datasources. Removing source_link was a good idea, but broke them :(
- The country field changes may have broken graphQL, but because refresh_datasources is also down. You’ll need to check manually after fixing the other issues.

shuesken commented:
I addressed some of your comments and got the tests to run again. GraphQL still works (although I have to rework the schema bit to include datasources).

Also fixed `ingest_airtable.py`. 

I can move to TimestampedModel for Brands as well, but figured that might be better for a second PR. Happy to put it in this one as well if you prefer them more complete.

RogerTangos commented:
You forgot to push ;)

RogerTangos reviewed (approved):
Approving.

The subclass() tests should really be its own function that tests at least two banks, but I gotta get this merged

The reason is that an older version of subclass defaulted to banktrack, so it's important to actually create two different data sources and test subclass on them both...
